### PR TITLE
Replay ZK logs using keeper-bench

### PR DESCRIPTION
**Pull request by antonio2368**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

to run it:

```sh
./utils/keeper-bench/keeper-bench --input-request-log zookeeper_log_all.tsv.zstd
```

if the zookeeper log doesn't start from the beginning, you can do a form of dry run which will just go through the requests in a file and generate a snapshot file based on it which will be the starting state for future runs

```sh
./utils/keeper-bench/keeper-bench --input-request-log zookeeper_log_all.tsv.zstd --setup-nodes-snapshot-path snapshots
```

snapshots is a directory in this case

the input file needs to have specific columns
I used this query
[pastila.nl/?02903a86/ea5db22316080e21f3ec20e4a8f9cb45#W8tcFYnWTztc5xWEowKWGw==](https://pastila.nl/?02903a86/ea5db22316080e21f3ec20e4a8f9cb45#W8tcFYnWTztc5xWEowKWGw==)
I just need to turn it into a query using clusterAllReplicas

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

---
### Modify your CI run:
**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Include tests (required builds will be added automatically):
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Unit tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with Analyzer
- [ ]  Add your option here

#### Exclude tests:
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with MSAN
- [ ]  All with UBSAN
- [ ]  All with Coverage
- [ ]  All with Aarch64
- [ ]  Add your option here

#### Extra options:
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ]  1
- [ ]  2
- [ ]  3
- [ ]  4

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit 95af46303ccaedeedf147801b04e5eaaac9004a7 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/62481/95af46303ccaedeedf147801b04e5eaaac9004a7/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: utils/keeper-bench/Runner.cpp [hunk 1]
```diff
@@ -328,9 +370,519 @@ bool Runner::tryPushRequestInteractively(Coordination::ZooKeeperRequestPtr && re
...
+            format,
+            *file_read_buf,
+            header_block,
+            context,
+            context->getSettingsRef().max_block_size,
+            format_settings,
+            1,
+            std::nullopt,
+            /*is_remote_fs*/ false,
+            DB::CompressionMethod::None,
```
Comment (alesapin) [inline utils/keeper-bench/Runner.cpp:564]: If we specify method here we can avoid `wrapReadBufferWithCompressionMethod`

**Review by alesapin — COMMENTED** [review]
🔥

File: utils/keeper-bench/Runner.cpp (hunk 1 above)
Comment (antonio2368) [reply utils/keeper-bench/Runner.cpp:572]: is there a benefit for that?
this seems more flexible

**Comment by antonio2368** [comment]
@alesapin it is ready :pray: 

to run it:
```
./utils/keeper-bench/keeper-bench --input-request-log zookeeper_log_all.tsv.zstd
```

if the zookeeper log doesn't start from the beginning, you can do a form of dry run which will just go through the requests in a file and generate a snapshot file based on it which will be the starting state for future runs

```
./utils/keeper-bench/keeper-bench --input-request-log zookeeper_log_all.tsv.zstd --setup-nodes-snapshot-path snapshots
```

`snapshots` is a directory in this case

the input file needs to have specific columns
I used this query
https://pastila.nl/?02903a86/ea5db22316080e21f3ec20e4a8f9cb45#W8tcFYnWTztc5xWEowKWGw==
I just need to turn it into a query using `clusterAllReplicas`

File: utils/keeper-bench/Runner.cpp [hunk 2]
```diff
@@ -328,9 +377,750 @@ bool Runner::tryPushRequestInteractively(Coordination::ZooKeeperRequestPtr && re
...
+                break;
+            }
+            case Coordination::OpNum::Check:
+            case Coordination::OpNum::CheckNotExists:
+            {
+                auto check_request = std::make_shared<Coordination::ZooKeeperCheckRequest>();
+                check_request->path = current_block->getPath(idx_in_block);
+                if (auto version = current_block->getVersion(idx_in_block))
+                {
+                    /// we just need to make sure that the request with version that need to fail, fail when replaying
```
Comment (alesapin) [inline utils/keeper-bench/Runner.cpp:659]: :fire:

File: utils/keeper-bench/Runner.cpp [hunk 3]
```diff
@@ -328,9 +377,750 @@ bool Runner::tryPushRequestInteractively(Coordination::ZooKeeperRequestPtr && re
...
+            auto & stats = request->isReadRequest() ? request_stats.read_requests : request_stats.write_requests;
+
+            stats.total.fetch_add(1, std::memory_order_relaxed);
+
+            if (expected_result)
+            {
+                if (*expected_result != response.error)
+                    stats.unexpected_results.fetch_add(1, std::memory_order_relaxed);
+
+                //if (*expected_result != response.error)
```
Comment (alesapin) [inline utils/keeper-bench/Runner.cpp:975]: Let's move to `ifdef`?

File: utils/keeper-bench/Runner.cpp [hunk 4]
```diff
@@ -458,8 +1248,232 @@ std::vector<std::shared_ptr<Coordination::ZooKeeper>> Runner::refreshConnections
...
+            children_span = children_span.subspan(0, children_span.size() - 1);
+        }
+        auto multi_promise = std::make_shared<std::promise<void>>();
+        auto multi_future = multi_promise->get_future();
+
+        auto multi_callback = [multi_promise] (const Coordination::MultiResponse &)
+        {
+            multi_promise->set_value();
+        };
+        zookeeper.multi(ops, multi_callback);
```
Comment (alesapin) [inline utils/keeper-bench/Runner.cpp:1305]: Was `multiAsync` supposed to be here?

**Review by alesapin — APPROVED** [review]
I'll try to use it before merge.

File: utils/keeper-bench/Runner.cpp (hunk 4 above)
Comment (antonio2368) [reply utils/keeper-bench/Runner.cpp:1305]: no, this is `Coordination::ZooKeeper` and not `zkutil::ZooKeeper` which is really obvious when the classes have same name :smile:

**Pull request merged**